yamux/src/frame/io.rs

```diff
@@ -152,66 +114,58 @@ impl<T: AsyncRead + AsyncWrite + Unpin> Sink<Frame<()>> for Io<T> {
 
 /// The stages of reading a new `Frame`.
 enum ReadState {
-    /// Initial reading state.
-    Init,
     /// Reading the frame header.
     Header {
         offset: usize,
         buffer: [u8; header::HEADER_SIZE],
     },
     /// Reading the frame body.
-    Body {
-        header: header::Header<()>,
-        offset: usize,
-        buffer: Vec<u8>,
-    },
+    Body { frame: Frame<Data>, offset: usize },
+}
+
+impl ReadState {
+    fn header() -> Self {
+        ReadState::Header {
+            offset: 0,
+            buffer: [0u8; header::HEADER_SIZE],
+        }
+    }
 }
 
 impl<T: AsyncRead + AsyncWrite + Unpin> Stream for Io<T> {
     type Item = Result<Frame<()>, FrameDecodeError>;
 
     fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
-        let mut this = &mut *self;
+        let this = &mut *self;
         loop {
             log::trace!("{}: read: {:?}", this.id, this.read_state);
             match this.read_state {
-                ReadState::Init => {
-                    this.read_state = ReadState::Header {
-                        offset: 0,
-                        buffer: [0; header::HEADER_SIZE],
-                    };
-                }
                 ReadState::Header {
                     ref mut offset,
                     ref mut buffer,
                 } => {
                     if *offset == header::HEADER_SIZE {
-                        let header = match header::decode(buffer) {
-                            Ok(hd) => hd,
-                            Err(e) => return Poll::Ready(Some(Err(e.into()))),
-                        };
+                        let frame = Frame::try_from_header_buffer(buffer)?;
```

Comment:
If someone send data frames with a large len field, Frame::try_from_header_buffer will call Frame::new and allocate large buffers for nothing because they will be dropped by the check on max_body_len. That's why I did introduce ensure_buffer_len() to resize the buffer after the check and prevent a potential denial of service attack.

Reply:
Good catch, we should probably just do the check inside then! I'll send a follow-up commit.